WordPress/src/main/java/org/wordpress/android/ui/posts/PrepublishingCategoriesViewModel.kt

```diff
@@ -0,0 +1,223 @@
+    private val networkUtilsWrapper: NetworkUtilsWrapper,
+    @Named(BG_THREAD) private val bgDispatcher: CoroutineDispatcher
+) : ScopedViewModel(bgDispatcher) {
+    private var isStarted = false
+    private lateinit var editPostRepository: EditPostRepository
+    private lateinit var siteModel: SiteModel
+    private var updateCategoriesJob: Job? = null
+
+    private val _navigateToHomeScreen = MutableLiveData<Event<Unit>>()
+    val navigateToHomeScreen: LiveData<Event<Unit>> = _navigateToHomeScreen
+
+    private val _navigateToAddCategoryScreen = MutableLiveData<Event<Unit>>()
+    val navigateToAddCategoryScreen: LiveData<Event<Unit>> = _navigateToAddCategoryScreen
+
+    private val _toolbarTitleUiState = MutableLiveData<UiString>()
+    val toolbarTitleUiState: LiveData<UiString> = _toolbarTitleUiState
+
+    private val _snackbarEvents = MediatorLiveData<Event<SnackbarMessageHolder>>()
+    val snackbarEvents: LiveData<Event<SnackbarMessageHolder>> = _snackbarEvents
+
+    private val _uiState: MutableLiveData<UiState> = MutableLiveData()
+    val uiState: LiveData<UiState> = _uiState
+
+    private val _refreshingUiState = MutableLiveData<SwipeToRefreshUiState>()
+    val refreshingUiState: LiveData<SwipeToRefreshUiState> = _refreshingUiState
+
+    fun start(editPostRepository: EditPostRepository, siteModel: SiteModel) {
+        this.editPostRepository = editPostRepository
+        this.siteModel = siteModel
+
+        if (isStarted) return
+        isStarted = true
+
+        init()
+    }
+
+    private fun init() {
+        setToolbarTitleUiState()
+
+        val siteCategories = getSiteCategories()
+        val postCategories = getPostCategories()
+        _uiState.value = ContentUiState(
+                siteCategories = siteCategories,
+                selectedCategoryIds = postCategories.toHashSet()
+        )
+    }
+
+    private fun setToolbarTitleUiState() {
+        _toolbarTitleUiState.postValue(UiStringRes(string.prepublishing_nudges_toolbar_title_categories))
+    }
+
+    fun onBackButtonClicked() {
+        _navigateToHomeScreen.postValue(Event(Unit))
+    }
+
+    fun onAddNewCategoryClicked() {
+        _navigateToAddCategoryScreen.postValue(Event(Unit))
+    }
+
+    override fun onCleared() {
+        super.onCleared()
+        updateCategoriesJob?.cancel()
+    }
+
+    fun updateCategories() {
+        if (!hasChanges()) return
+
+        if (!networkUtilsWrapper.isNetworkAvailable()) {
+            _snackbarEvents.postValue(
+                    Event(SnackbarMessageHolder(UiStringRes(string.no_network_message)))
+            )
+            return
+        }
+
+        val categoryList = (uiState.value as? ContentUiState)?.selectedCategoryIds ?: return
+        updateCategoriesJob?.cancel()
+        updateCategoriesJob = launch(bgDispatcher) {
+            updatePostCategoriesUseCase.updateCategories(categoryList.toList(), editPostRepository)
+        }
+        trackCategoriesChangedEvent()
```

Comment:
Ugh, I suppose I can move it to the home fragment since we don't wait around for success or failure. The same would be true for add category - the capture would have to occur in the select logic. Bleh. Noted in the issue #13066.